### PR TITLE
SLING-11504 BundleResource/JcrNodeResource getResourceType fallback consistency 

### PR DESCRIPTION
**Pull request by michalcukierman**

### CODE REVIEW

File: src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java [hunk 1]
```diff
@@ -138,17 +135,19 @@ public BundleResource(final ResourceResolver resourceResolver,
...
                     final URL url = this.cache.getEntry(propsPath);
                     if (url != null) {
-                        final JsonObject obj = Json.createReader(url.openStream()).readObject();
-                        for(final Map.Entry<String, JsonValue> entry : obj.entrySet()) {
-                            final Object value = getValue(entry.getValue(), true);
-                            if ( value != null ) {
-                                if ( value instanceof Map ) {
-                                    if ( children == null ) {
-                                        children = new HashMap<>();
+                        try (JsonReader reader = Json.createReader(url.openStream())) {
```
Comment (joerghoh) [inline src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java:138]: If this is a general refactoring, please raise a dedicated PR for it. Mixing multiple aspects in a single PR makes it harder to review.

File: src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java [hunk 2]
```diff
@@ -186,7 +185,7 @@ Resource getChildResource(final String path) {
         return result;
     }
 
-    private static Object getValue(final JsonValue value, final boolean topLevel) {
+    private static Object getValue(final JsonValue value) {
```
Comment (joerghoh) [inline src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java:188]: same here, seems like a cleanup unrelated to the PR

File: src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java [hunk 3]
```diff
@@ -229,7 +228,11 @@ public String getPath() {
 
     @Override
     public String getResourceType() {
-        return this.valueMap.get(ResourceResolver.PROPERTY_RESOURCE_TYPE, String.class);
+        String resourceType = this.valueMap.get(ResourceResolver.PROPERTY_RESOURCE_TYPE, String.class);
+        if ( resourceType == null ) {
+            resourceType = this.isFolder ? NT_FOLDER : NT_FILE;
```
Comment (joerghoh) [inline src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java:233]: food for thoughts: Why do we use JCR nodetypes here? Wouldn't it be possible to use some sling "default resourcetypes"?

**Review by joerghoh — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java (hunk 1 above)
Comment (michalcukierman) [reply src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java]: reverted

File: src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java (hunk 2 above)
Comment (michalcukierman) [reply src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java]: reverted

File: src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java (hunk 3 above)
Comment (michalcukierman) [reply src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java:230]: I've checked DefaultGetServlets and it looks like this type is not used by default for rendering the content. For directories, we fall into this condition: https://github.com/apache/sling-org-apache-sling-servlets-get/blob/org.apache.sling.servlets.get-2.1.44/src/main/java/org/apache/sling/servlets/get/impl/helpers/StreamRenderer.java#L131 , for files we pass control back to BundleResource: https://github.com/apache/sling-org-apache-sling-servlets-get/blob/org.apache.sling.servlets.get-2.1.44/src/main/java/org/apache/sling/servlets/get/impl/helpers/StreamRenderer.java#L157

Comment (michalcukierman) [reply src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java:230]: @joerghoh are you thinking about default resource types across all resource providers? Lets say: `sling/default/file`, `sling/deafult/folder` or `sling/default/unstructure`? 

Or rather to introduce something specific to BundleResource, like we have it in ServletResource superType [here](https://github.com/apache/sling-org-apache-sling-servlets-resolver/blob/0e3cdecc9cf4f88a740ba5fc86dec8d1dc9e2c17/src/main/java/org/apache/sling/servlets/resolver/internal/resource/ServletResource.java#L40)?

Comment (michalcukierman) [reply src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java:230]: The following implementation (local default resource types) seems to be working:
```
    @Override
    public String getResourceType() {
        String resourceType = this.valueMap.get(ResourceResolver.PROPERTY_RESOURCE_TYPE, String.class);
        if ( resourceType == null ) {
            resourceType = this.isFolder ? "sling/bundle/folder" : "sling/bundle/file";
        }
        return resourceType;
    }
```

Comment (kwin) [reply src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java:230]: For me using completely JCR agnostic resource types should be tracked in a follow-up PR. IMHO this PR is only about removing the resource type from the valueMap.

**Review by kwin — APPROVED** [review]
LGTM now

File: src/test/java/org/apache/sling/bundleresource/impl/BundleResourceTest.java [hunk 4]
```diff
@@ -111,7 +111,7 @@ void addContent(BundleResourceCache cache, String path, String content) throws I
         assertEquals(JcrConstants.NT_FILE, rsrc.getResourceType());
         assertNull(rsrc.getResourceSuperType());
         final ValueMap vm = rsrc.getValueMap();
-        assertEquals(JcrConstants.NT_FILE, vm.get(ResourceResolver.PROPERTY_RESOURCE_TYPE, String.class));
+        assertNull(vm.get(ResourceResolver.PROPERTY_RESOURCE_TYPE, String.class));
         assertEquals("foo", vm.get("test", String.class));
```
Comment (kwin) [inline src/test/java/org/apache/sling/bundleresource/impl/BundleResourceTest.java:115]: Another test case for JSON resources would be nice, which asserts that explicit "sling:property" keys from the JSON also appear in the value map. @michalcukierman Can you extend an existing test or add a new one to cover that?

Comment (michalcukierman) [reply src/test/java/org/apache/sling/bundleresource/impl/BundleResourceTest.java:115]: Tests added

File: src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java (hunk 3 above)
Comment (michalcukierman) [reply src/main/java/org/apache/sling/bundleresource/impl/BundleResource.java:241]: Agree, out of the scope of this PR.

File: src/main/java/org/apache/sling/bundleresource/impl/PathMapping.java [hunk 5]
```diff
@@ -44,18 +46,20 @@ public static PathMapping[] getRoots(final String rootList) {
             final String resourceRoot = entry.getValue();
             final String pathDirective = entry.getDirectiveValue(DIR_PATH);
             final String expandDirective = entry.getDirectiveValue(DIR_JSON);
+            final boolean skipSettingResourceTypeProperty = Boolean.parseBoolean(entry.getDirectiveValue(SKIP_SETTING_RESOURCE_TYPE_FLAG));
```
Comment (kwin) [inline src/main/java/org/apache/sling/bundleresource/impl/PathMapping.java:49]: Maybe "hideResourceTypeProperties" is the better name (and an according constant name value).

**Review by joerghoh — APPROVED** [review]

File: src/main/java/org/apache/sling/bundleresource/impl/PathMapping.java (hunk 5 above)
Comment (michalcukierman) [reply src/main/java/org/apache/sling/bundleresource/impl/PathMapping.java:49]: It was proposed by @cziegeler in the [jira comment](https://issues.apache.org/jira/browse/SLING-11504?focusedCommentId=17574690&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17574690)

I think that skipSettingResourceTypeProperty describes the actual flag behavior. Hiding a property may be understood in different ways.

**Pull request closed without merging**